### PR TITLE
Complete GEPA launch function

### PR DESCRIPTION
**Pull request by anndvision**
This PR implements the complete GEPA launch function with full Pareto frontier management and validation. The implementation includes:

- Completion of the GEPA launch function to seed, maintain, and update a Pareto frontier across iterations
- Introduction of `GEPAVariant` struct for better type safety and code organization
- Dataset creation now returns datapoint IDs for Pareto frontier initialization
- Validation of examples with detailed error reporting
- Error handling with explicit documentation of continue choices when evaluations fail
- Extensive E2E test coverage for both Chat and JSON functions

**Dependencies:** This PR depends on #4739 (Pareto filtering) and #4845 (mutate function), both of which have been merged.

---

### 1. GEPA Launch Function (`mod.rs`)

**New Architecture:**
- **Pareto Frontier Initialization:** Seeds the frontier with validation datapoint IDs and evaluates all initial variants on the validation set
- **Frequency-based Parent Sampling:** Selects parent variants proportionally to their instance-wise Pareto membership counts
- **Minibatch Validation Gating:** Only evaluates children on the full validation set if they Pareto-dominate the parent on the minibatch
- **Incremental Frontier Updates:** Updates Pareto frontier throughout iterations, maintaining only non-dominated variants
- **Original Variant Filtering:** Returns only newly created variants by filtering out the original baseline variants

**Key Implementation Details:**
- Divides concurrency across variants to prevent max_concurrency² explosion
- Handles evaluation failures gracefully with `continue` statements and detailed logging
- Adds TODO for semaphore sharing with `run_evaluation_core_streaming` (#4914)
- Cleans up temporary datasets at the end of optimization
- Uses `join_all` for parallel evaluation of initial variants, collecting errors instead of failing fast

**Error Handling Philosophy:**
- **Initial variant evaluation:** Collect all errors, proceed if at least one succeeds (lines 155-224)
- **Per-iteration operations:** Skip iteration on failure, continue optimization (lines 321-323, 418-419)
- Detailed tracing logs explain why iterations are skipped
- Explanatory comments clarify why different error handling patterns are appropriate for different contexts

### 2. Type Safety Enhancement: GEPAVariant Struct (`mod.rs`)

Introduced dedicated struct for better type safety:
```rust
pub struct GEPAVariant {
    pub name: VariantName,
    pub config: UninitializedChatCompletionConfig,
}
```

**Benefits:**
- Clearer function signatures (no more HashMap single-entry patterns)
- Better code documentation
- Reduced error-prone tuple manipulation
- Consistent representation throughout the codebase

### 3. Dataset Creation Enhancement (`evaluate.rs`)

**Key Change:** `create_evaluation_dataset` now returns `CreateDatapointsResponse` containing datapoint IDs

**Impact:**
- Enables Pareto frontier initialization with all datapoint IDs
- Provides proper tracking of which datapoints are in the validation set
- Critical for instance-wise Pareto set computation with caching

### 4. Mutate Function Updates (`mutate.rs`)

- refactor `parent` argument `HashMap<&String, &UninitializedChatCompletionConfig>` -> `&GEPAVariant`

### 5. Pareto Frontier Logic (`pareto.rs`)

- Returns `GEPAVariant` from `sample_by_frequency()` instead of HashMap

### 6. Validation Enhancements (`validate.rs`)

**Renamed Function:**
- `initialize_pareto_frontier` → `get_uninitialized_variant_configs` (more descriptive)

### 7. Comprehensive E2E Tests (`tests/e2e/gepa/mod.rs` - 773 lines changed)

**Major Test Expansion:**

Both Chat and JSON tests now validate complete GEPA iteration flow:
- Evaluate ALL initial variants on validation dataset (not just one)
- Seed Pareto frontier with initial candidates
- Sample parent from frontier using `sample_by_frequency()`
- Create separate minibatch/training dataset
- Evaluate parent on minibatch
- Run analysis on minibatch results
- Mutate to create child variant
- Evaluate child on minibatch
- Test `is_improvement()` function (compare parent vs child)
- Conditionally evaluate child on validation set if improved
- Update Pareto frontier with child
- Verify frontier state (non-empty, can sample)
- Clean up both validation and training datasets

**Test Infrastructure:**
- Added `TEST_CLICKHOUSE_WAIT_MS` constant (500ms) for consistent wait times
- Added imports: `Candidate`, `is_improvement`, `VariantName`
- Fixed Pareto frontier assertions to handle dominated variant filtering

---

### CODE REVIEW

File: tensorzero-optimizers/tests/e2e/gepa/mod.rs
[the remark this answers is not in the document]
Comment (anndvision) [reply tensorzero-optimizers/tests/e2e/gepa/mod.rs]: Good point! I opened PR #4929 to address this concern by refactoring these monolithic tests into a modular structure.
